### DISCUSSION TITLE
Rebuild vs merge for post-1.21

### DISCUSSION THREAD

**Discussion by rnicoll** [dogecoin/dogecoin#2273] [General]
For Dogecoin Core 1.10, 1.14 and 1.21 we have specifically started from the relevant version of Bitcoin Core (0.10, 0.14 and 0.21 respectively) then "Dogeified" it by re-introducing all of the changes that make it Dogecoin. This is done because historically there are fewer changes to make Bitcoin Core into Dogecoin Core, than have happened to Bitcoin Core between Dogecoin Core releases.

Looking forward, we may need to consider if this is the most practical option. This is entwined in how often we do releases (and Dogecoin intentionally does fewer releases so people don't have to do extensive compatibility testing as often, but instead have only minor revisions applied). However, if we were to try updating more often, we may need to consider that merging changes from Bitcoin Core in might be more practical, especially given we are introducing more significant divergence in the UI code (i.e. adding functionality to add nodes via the GUI).

What do others think?

**Comment by patricklodder** [comment]
Personally, I have never liked the rebuilds. Besides that we disincentivize innovation on our own code, the rebuilds are extremely risky and both 1.10 and 1.14 introduced problems (resp. memory usage and network issues) that could have been prevented by doing smaller, incremental releases and only doing the entire rebuilds when it is impossible to enable a feature we need. I hope we can improve this with 1.21 (more, better testing) but by doing a rebuild, it kind of means we increase mitigations (= working harder) and not prevention (= working smarter). 

Instead, keeping dependencies up-to-date, back porting bug fixes and having deeper discussions about soft forks independently from bitcoin discussions that have clear developer bias... would be great. 

I would not want to exclude the option to do a full rebuild though. It may still sometimes be beneficial, i.e. when there are a lot of changes in bitcoin/bitcoin that enable a feature we want (i.e. if we'd want taproot-like functionality, doing a rebuild instead of a cherry pick onto 1.14 would have made a lot of sense) but always depending on rebuilds for major releases is suboptimal. 

For example, we're currently discussing soft forks that are available in 1.14, yet we're rebuilding 1.21. It feels like a lot of unnecessary effort is being spent, introducing bigger risks, even while we could have probably back-ported every beneficial fix.

**Reply by rnicoll** [reply]
Awesome; yeah, I think 1.10 made sense, I hope 1.21 makes sense, but hopefully this will be the last rebuild and from here we can merge meaningfully to a much cleaner baseline.

**Reply by micaelmalta** [reply]
Maybe we should think on a Plugin system instead of having to backport everytime Bitcoin commits: is something eve possible?

**Reply by rnicoll** [reply]
Unfortunately would probably require we make Bitcoin Core pluggable, which I suspect isn't likely to be something they'd accept.

**Reply by ReverseControl** [reply]
I would favor picking a baseline, building ourselves from there, back port progress as we go along and own our code base entirely so we may innovate. Plus, design decisions will be ours, instead of using whatever the bitcoin community agreed makes sense for them -- it might not make sense for us. We will happily pick from their improvements if they make sense for dogecoin.

**Reply by chromatic** [reply]
I have mixed feelings here, having worked around some Bitcoin implementation decision in PR #2698 and looking at others for GH #1313. Any larger refactorings to make implementing things in Dogecoin easier run the risk of making backports from Bitcoin more difficult. The cleanest way to bridge that gap is to attempt to contribute upstream, then eventually cherry pick those changes if accepted.

Maybe the question is about what kind of backports we value and what kind of Dogecoin-only improvements we value.
